akka-actor/src/main/scala/akka/io/dns/CachePolicy.scala

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.io.dns
+
+import akka.annotation.ApiMayChange
+
+import scala.concurrent.duration.{ Duration, FiniteDuration, _ }
+
+object CachePolicy {
+
+  @ApiMayChange
+  sealed trait CachePolicy
+  @ApiMayChange
+  case object Never extends CachePolicy
+  @ApiMayChange
+  case object Forever extends CachePolicy
+  @ApiMayChange
+  sealed abstract case class Ttl(value: FiniteDuration) extends CachePolicy {
```

Comment:
why is this `abstract` (case classes should be final in my opinion)
if you want to enforce creation from the factory methods in the companion you can make the constructor private `final case class Ttl private (value: FiniteDuration)`

Reply:
IIRC, making this `abstract` prevents the creation of `copy` and makes it impossible to break the invariant we want to enforce (`value > Duration.zero`) which is on the `fromPositive` factory method.

Reply:
doesn't prevent new instances to be created from java

that invariant should be checked in the constructor instead

Reply:
:+1: for making the constructor private (good catch)
:-1: for throwing in the constructor: throw in the obviously-named factory method and disallow the copy method

Reply:
(more info: https://twitter.com/tpolecat/status/788215281046917120)

Reply:
private constructor would not be private from java (I think) so if we really care about the invariant it should be in the constructor

the obviously-named factory method will be in the stack trace

Reply:
Ah yes you're right!  In order to allow the companion object to access it it can't be private.  It could be java package private, not sure why it isn't...

So move the check into the constructor
Just call new in the factory method
Keep abstract to elide generating copy.

Reply:
Using `abstract` is trivially easier to add than re-implementing unapply, equals, hashCode and toString.  😄

But you're right that you can't maintain `unapply`'s binary compatibility with `case class`...

Reply:
Thanks again @patriknw! We overlooked the java use cases in earlier iterations. I've just pushed a version that I think reflects all the suggested changes and includes `apply` as a bonus.